airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -0,0 +1,133 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import asyncio
+import logging
+from datetime import datetime
+from functools import cache
+from typing import Any, Iterable, List, Mapping, Optional, Union
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.file_based.exceptions import MissingSchemaError, RecordParseError, SchemaInferenceError
+from airbyte_cdk.sources.file_based.remote_file import FileType, RemoteFile
+from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas, type_mapping_to_jsonschema
+from airbyte_cdk.sources.file_based.schema_validation_policies import record_passes_validation_policy
+from airbyte_cdk.sources.file_based.stream import AbstractFileBasedStream
+from airbyte_cdk.sources.utils.record_helper import stream_data_to_airbyte_message
+
+
+class DefaultFileBasedStream(AbstractFileBasedStream):
+    """
+    The default file-based stream.
+    """
+
+    @property
+    def primary_key(self) -> Optional[Union[str, List[str], List[List[str]]]]:
+        return self.config.primary_key
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
```

Comment:
reading through this with fresh eyes...

I think we should use the stream_slice to decide which files to read.

The flow I'd expect is
1. Given a stream_state, `stream_slices` generates partitions/slices to process (probably 1 file per slice to leverage the default checkpointing behavior
2. `read_records` uses the stream_slice instead of the stream_state to know which file to read. 

Using the stream_slice instead of the state will allow us to parallelize the files and streams by pushing all the parallelization logic to the source

Reply:
agree w alex, stream slice should contain all the info needed to read the records in a file

Reply:
I think the reasoning is: slices are a function of state, so we should be able to encode anything we need from the state into the slice

For context I think we're trying to push in the direction of having slices completely describe requests, more [here](https://www.loom.com/share/100bd529cc8f49c3877cfaa235f91ae6)

Reply:
^+1
Would be interesting to get some numbers to confirm, but this is also how most low-code connectors work. We usually only need to reference the stream_slice and not the stream_state in the outgoing request

Reply:
👍 got it, makes sense to me. I've updated the signature here and in the downstream functions.

`stream_state` is required by the abstractmethod `read_records` from the `Stream` interface, which we're inheriting from so I've left that in.